### PR TITLE
Late ack record container dynamic

### PR DESCRIPTION
**Pull request by lbudai**
This branch implements the DynamicLateAckRecordContainer (and depends on #2787 ).

( PR #2772 also contains these, but can be reviewed independently, if you review it & merge, then I'll rebase PR #2772 ).

### CODE REVIEW

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: lib/ack-tracker/late_ack_record_container_dynamic.c [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+  if (n == self->size)
+    {
+      g_list_free_full(self->head, g_free);
+      self->head = NULL;
+      self->size = 0;
+      return;
+    }
+
+  GList *delete_head = self->head;
+  GList *last = g_list_nth(self->head, n-1);
```
Comment (MrAnno) [inline lib/ack-tracker/late_ack_record_container_dynamic.c]: Do you want to assert/handle `n > self->size` here too?

File: lib/ack-tracker/late_ack_record_container_dynamic.c [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+
+static void
+_drop(LateAckRecordContainer *s, gsize n)
+{
+  DynamicLateAckRecordContainer *self = (DynamicLateAckRecordContainer *) s;
+
+  if (n == self->size)
+    {
+      g_list_free_full(self->head, g_free);
+      self->head = NULL;
```
Comment (MrAnno) [inline lib/ack-tracker/late_ack_record_container_dynamic.c]: `self->tail` should be reset too.

File: lib/ack-tracker/late_ack_record_container_dynamic.c [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+  if (self->pending)
+    {
+      late_ack_record_destroy(self->pending);
+      g_free(self->pending);
+      self->pending = NULL;
+    }
+
+  g_list_free_full(self->head, g_free);
+  self->head = NULL;
```
Comment (MrAnno) [inline lib/ack-tracker/late_ack_record_container_dynamic.c]: It doesn't really matter, but `self->tail` could be nulled as well.

File: lib/ack-tracker/late_ack_record_container_dynamic.c (hunk 1 above)
Comment (MrAnno) [reply lib/ack-tracker/late_ack_record_container_dynamic.c]: The last commits fixes this.

File: lib/ack-tracker/late_ack_record_container_dynamic.c [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+static LateAckRecord *
+_at(const LateAckRecordContainer *s, gsize idx)
+{
+  DynamicLateAckRecordContainer *self = (DynamicLateAckRecordContainer *) s;
+
+  return (LateAckRecord *) g_list_nth(self->head, idx)->data;
+}
+
+static void
+_free(LateAckRecordContainer *s)
```
[the remark this answers is not in the document]
Comment (lbudai) [reply lib/ack-tracker/late_ack_record_container_dynamic.c]: indeed
fixed

File: lib/ack-tracker/late_ack_record_container_dynamic.c (hunk 2 above)
Comment (lbudai) [reply lib/ack-tracker/late_ack_record_container_dynamic.c]: fixed

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Pull request merged**